jax/_src/pallas/mosaic/interpret.py

```diff
@@ -0,0 +1,694 @@
+CostEstimate = pallas_core.CostEstimate
+
+
+@dataclasses.dataclass(frozen=True)
+class SharedMemory:
+  # (memory_space, buffer_id, device_id) -> NumPy array
+  mem: dict = dataclasses.field(default_factory=dict)
+
+  # semaphore_id -> Semaphore
+  sem: dict = dataclasses.field(default_factory=dict)
+
+  lock: threading.Lock = dataclasses.field(default_factory=threading.Lock)
+
+  next_buffer_id: dict = dataclasses.field(
+    default_factory=lambda: collections.defaultdict(lambda: 100))
+  next_semaphore_id: dict = dataclasses.field(
+    default_factory=lambda: collections.defaultdict(lambda: 2000))
+
+shared_memory = SharedMemory()
+
+class Semaphore:
+  def __init__(self, num_devices=1):
+    self.cv = threading.Condition()
+
+    # TODO(jburnim): Index counts by a single integer device ID, instead of
+    # mesh coordinates.
+    self.counts = collections.defaultdict(int)
+
+  def signal(self, inc, device_id):
+    device_id = tuple(int(x) for x in device_id)
+    with self.cv:
+      self.counts[device_id] += inc
+      self.cv.notify_all()
+
+  def wait(self, value, device_id):
+    device_id = tuple(int(x) for x in device_id)
+    with self.cv:
+      while self.counts[device_id] < value:
+        self.cv.wait()
+      self.counts[device_id] -= value
+
+def _allocate_buffer(device_id, memory_space, val):
+  device_id = tuple(map(int, device_id))
+  memory_space = TPU_MEMORY_SPACE_NAMES[int(memory_space)]
+  val = np.array(val)
+
+  with shared_memory.lock:
+    buffer_id = shared_memory.next_buffer_id[device_id]
+    shared_memory.next_buffer_id[device_id] = buffer_id + 1
+    shared_memory.mem[(memory_space, buffer_id, device_id)] = val
+
+  return np.int16(buffer_id)
+
+def _allocate_semaphore(device_id, shape):
+  device_id = tuple(map(int, device_id))
+  shape = tuple(map(int, shape))
+  num_semaphores = math.prod(shape)
+
+  with shared_memory.lock:
+    semaphore_id = shared_memory.next_semaphore_id[device_id]
+    shared_memory.next_semaphore_id[device_id] = semaphore_id + num_semaphores
+    for i in range(semaphore_id, semaphore_id + num_semaphores):
+      if not i in shared_memory.sem:
+        shared_memory.sem[i] = Semaphore()
+
+  return np.int16(
+      range(semaphore_id, semaphore_id + num_semaphores)
+  ).reshape(shape)
+
+
+TPU_MEMORY_SPACE_IDXS = {
+    v: i for i, v in enumerate(mosaic_core.TPUMemorySpace)}
+TPU_MEMORY_SPACE_NAMES = {
+    i: v.value for i, v in enumerate(mosaic_core.TPUMemorySpace)}
+
+def get_barrier_semaphore(device_id, collective_id):
+  device_id = tuple(map(int, device_id))
+  collective_id = int(collective_id)
+
+  # TODO(jburnim): Check/fix so that IDs for barrier semaphores do not conflict
+  # with IDs for regular or DMA semaphores.  (For example, store them in a
+  # different table.)
+  with shared_memory.lock:
+    semaphore_id = collective_id
+    if not semaphore_id in shared_memory.sem:
+      shared_memory.sem[semaphore_id] = Semaphore()
+
+  return np.int16(semaphore_id)
```

Comment:
Why does this need to be int16?

Reply:
My thinking was that it might be useful to match SEMAPHORE_INTERPRET_DTYPE in `pallas/core.py`:
```python
# Datatype for semaphore values in interpret mode.
# For now, we choose a relatively uncommon datatype (i16) so it is more easily
# identifiable in kernels.
# TODO(justinfu): Handle semaphores with a custom extended dtype.
SEMAPHORE_INTERPRET_DTYPE = jnp.int16
```

I'll add a comment noting this.

(I'm also happy to switch this dtype if you think int32 or something else makes more sense.)